samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs

```diff
@@ -212,27 +215,116 @@ public async Task<string> AcquireExternalInformationAsync(SKContext context)
+                planResult = this.OptimizeOpenApiSkillJson(planResult, relatedInformationTokenLimit, plan);
+            }
 
             string informationText = $"[START RELATED INFORMATION]\n{planResult.Trim()}\n[END RELATED INFORMATION]\n";
 
             // Adjust the token limit using the number of tokens in the information text.
-            int tokenLimit = int.Parse(context["tokenLimit"], new NumberFormatInfo());
             tokenLimit -= Utilities.TokenCount(informationText);
             context.Variables.Set("tokenLimit", tokenLimit.ToString(new NumberFormatInfo()));
 
             return informationText;
         }
-        else
+
+        return string.Empty;
+    }
+
+    /// <summary>
+    /// Try to optimize json from the planner response
+    /// based on token limit
+    /// </summary>
+    private string OptimizeOpenApiSkillJson(string jsonContent, int tokenLimit, Plan plan)
+    {
+        int jsonTokenLimit = (int)(tokenLimit * this._promptSettings.RelatedInformationContextWeight);
+
+        // Remove all new line characters + leading and trailing white space
+        jsonContent = Regex.Replace(jsonContent.Trim(), @"[\n\r]", string.Empty);
+        var document = JsonDocument.Parse(jsonContent);
+        string lastSkillInvoked = plan.Steps[^1].SkillName;
+        string lastSkillOperationInvoked = plan.Steps[^1].Name;
+
+        // Check if the last skill invoked was JiraSkill and deserialize the JSON content accordingly
+        if (string.Equals(lastSkillInvoked, "JiraSkill", StringComparison.Ordinal))
         {
-            return string.Empty;
+            Type skillResponseType = typeof(IssueResponse); // reasonable assumption of type so value isnt null
+            if (lastSkillOperationInvoked == "GetIssue")
+            {
+                skillResponseType = document.RootElement.ValueKind == JsonValueKind.Array ? typeof(IssueResponse[]) : typeof(IssueResponse);
+            }
+            //else if (lastSkillOperationInvoked == "AddComment")
+            //{
+            //    skillResponseType = document.RootElement.ValueKind == JsonValueKind.Array ? typeof(CommentResponse[]) : typeof(CommentResponse);
+            //}
+
+            // Deserializing limits the json content to only the fields defined in the GitHubSkill/Model classes
+            var skillResponse = JsonSerializer.Deserialize(jsonContent, skillResponseType);
+            jsonContent = skillResponse != null ? JsonSerializer.Serialize(skillResponse) : string.Empty;
+            document = JsonDocument.Parse(jsonContent);
+        }
+
+        int jsonContentTokenCount = Utilities.TokenCount(jsonContent);
+
+        // Return the JSON content if it does not exceed the token limit
+        if (jsonContentTokenCount < jsonTokenLimit)
+        {
+            return jsonContent;
+        }
+
+        List<object> itemList = new();
+
+        // Summary (List) Object
+        if (document.RootElement.ValueKind == JsonValueKind.Array)
```

Comment:
Could you add more explanation of what the strategy is here? Something like "To stay within token limits, attempt to truncate lists of results..."